### PR TITLE
Complete rework of our CI

### PR DESCRIPTION
**Pull request by AnotherDaniel**
This is to address #51, and will be a multi-stage PR (a few individual commits with change sets building up the whole picture).

Step 1 - enable cargo-deny, adding deny.toml, and changing a build.rs dep for avoid strange licensed crates in our dep tree

Step 2 - remove Cargo.lock from our .gitignore and add a couple of template files to enable cargo-about

Step 3 - introduce the all-new github workflows

### CODE REVIEW

File: .github/workflows/x-build.yaml
Comment (PLeVasseur) [inline .github/workflows/x-build.yaml:41]: :+1: 

Was about to ask for this yesterday on the issue, but you beat me to the punch :slightly_smiling_face:

**Comment by PLeVasseur** [comment]
Hey @AnotherDaniel -- opened an issue for adding cross-compiling: #73 

Do you think we should tackle that as a part of this PR or a new one after this one merges?

**Comment by AnotherDaniel** [comment]
God this workspace setup with locked-in required workflows is driving up the wall in this context here. Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/

**Comment by PLeVasseur** [comment]
> Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/

I dunno if commiters get admin rights. If so => @sophokles73, if not => @stevenhartley

**Comment by stevenhartley** [comment]
> > Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/
> 
> I dunno if committers get admin rights. If so => @sophokles73, if not => @stevenhartley

Hi @AnotherDaniel , that is me and I think anyone else. We use branch protection rules that require certain workflows to be successful to run. These rules are turned on/off using otterdog in the https://github.com/eclipse-uprotocol/.eclipsefdn project. As you can see in https://github.com/eclipse-uprotocol/.eclipsefdn/blob/main/otterdog/eclipse-uprotocol.jsonnet#L355 we require three workflows to be successful so if you change the names of the workflows (or delete them), the branch protection rules will fail. We can temporarily turn them off, fix the workflows, then turn them back on if you like? It is a 10 min change for me, let me know if you want me to do that for your project. You can also fork this project, make the change, then create a PR to merge it back in.

**Comment by AnotherDaniel** [comment]
@stevenhartley then I think the path forward would be to turn off these rules, merge this PR and turn them on again?

Would work for me, I can't do any of these things but would pick things up once these workflows are active.

**Comment by stevenhartley** [comment]
Change is pending, once merge you only have to retrigger the change by pushing something (black space or something): https://github.com/eclipse-uprotocol/.eclipsefdn/pull/46

**Review by stevenhartley — APPROVED** [review]

**Comment by stevenhartley** [comment]
@AnotherDaniel it is all passing and looking great, please let me know if it is ready to merge or if you would like others to chime in as well.

**Comment by AnotherDaniel** [comment]
I think this can be merged- already did a review with @sophokles73 earlier today, and this is in all dimensions an improvement to what we had before. More goodness can surely be added, but in subsequent PRs...

File: .github/workflows/check.yaml
Comment (PLeVasseur) [inline .github/workflows/check.yaml:2]: Update copyright to 2024?

**Review by PLeVasseur — APPROVED** [review]
Impressive stuff! Thanks @AnotherDaniel :slightly_smiling_face: 

I must admit, I am not as knowledgeable as you on the CI, but this looks comprehensive

File: deny.toml
Comment (sophokles73) [inline deny.toml:2]: Since you have added this file newly in 2024 I would suggest to change this accordingly.

File: about.hbs
Comment (sophokles73) [inline about.hbs:75]: IMHO this line deserves an EOL :-)

File: about.toml
Comment (sophokles73) [inline about.toml:19]: IMHO this line deserves an EOL :-)

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:1]: IMHO we do not need this line as the file name alreay suggests that contains documentation and the fact that we renamed the repository multiple times already also makes it questionable if it is a good idea to add the repo name here again ;-)

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:15]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:21]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:27]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:33]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:36]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:38]: how about linking to the file?

File: .github/workflows/README.md
Comment (sophokles73) [inline .github/workflows/README.md:39]: how about linking to the file?

File: .github/workflows/check.yaml
Comment (sophokles73) [inline .github/workflows/check.yaml:150]: you might want to remove the whitespace from the last line ...

File: .github/workflows/coverage.yaml
Comment (sophokles73) [inline .github/workflows/coverage.yaml:2]: 2024?

File: .github/workflows/coverage.yaml
Comment (sophokles73) [inline .github/workflows/coverage.yaml:81]: add an EOL?

File: .github/workflows/license-report.yaml
Comment (sophokles73) [inline .github/workflows/license-report.yaml:2]: 2024?

File: .github/workflows/license-report.yaml
Comment (sophokles73) [inline .github/workflows/license-report.yaml:57]: add an EOL?

File: .github/workflows/nightly.yaml
Comment (sophokles73) [inline .github/workflows/nightly.yaml:2]: 2024?

File: .github/workflows/release.yaml
Comment (sophokles73) [inline .github/workflows/release.yaml:2]: 2024?

File: .github/workflows/release.yaml
Comment (sophokles73) [inline .github/workflows/release.yaml:177]: you might want to remove the whitespace from the last line

File: .github/workflows/test-featurematrix.yaml
Comment (sophokles73) [inline .github/workflows/test-featurematrix.yaml:2]: 2024?

File: .github/workflows/x-build.yaml
Comment (sophokles73) [inline .github/workflows/x-build.yaml:2]: 2024?

**Comment by AnotherDaniel** [comment]
All review comments implemented

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by AnotherDaniel** [comment]
Ah, merge it as is - commit history is reasonably clear I'd say.

**Pull request merged**
